File: src/main/java/client/OSM/OSMMap.java
package OSM;

import javafx.concurrent.Worker;
import javafx.event.EventHandler;
import javafx.scene.Parent;
import javafx.scene.web.WebEngine;
import javafx.scene.web.WebView;
import netscape.javascript.JSObject;

public class OSMMap extends Parent {


    private double curLat;
    private double curLong;
    private double curZoom;

    public double getCurZeroLat() {
        return curZeroLat;
    }

    public double getCurZeroLong() {
        return curZeroLong;
    }

    private double curZeroLat;
    private double curZeroLong;
    private JSObject doc;
    private EventHandler<OSMMapEvent> onMapLatLngZoomChanged;
    private WebView webView;
    private WebEngine webEngine;
    private boolean ready;

    public OSMMap(){
        initMap();
        initCommunication();
    }
    public double getCurLat() {
        return curLat;
    }

    public double getCurLong() {
        return curLong;
    }

    public double getCurZoom() {
        return curZoom;
    }

    private void initMap()
    {
        webView = new WebView();
        webEngine = webView.getEngine();
        webEngine.setJavaScriptEnabled(true);
        webEngine.load(getClass().getResource("/resources/evtTest.html").toExternalForm());
        getChildren().add(webView);
        ready = false;
        webEngine.getLoadWorker().stateProperty().addListener((observableValue, oldState, newState) -> {
            if (newState == Worker.State.SUCCEEDED) {
                ready = true;
            }
        });
    }
    private void initCommunication() {
        webEngine.getLoadWorker().stateProperty().addListener((observableValue, oldState, newState) -> {
            if (newState == Worker.State.SUCCEEDED)
            {
                doc = (JSObject) webEngine.executeScript("window");
                doc.setMember("app", OSMMap.this);
            }
        });
    }
    private void invokeJS(final String str) {
        webEngine.executeScript(str);
        if(ready) {
            doc.eval(str);
        }
        else {
            webEngine.getLoadWorker().stateProperty().addListener((observableValue, oldState, newState) -> {
                if (newState == Worker.State.SUCCEEDED) {
                    doc.eval(str);
                }
            });
        }
    }
    public void handle(double lat, double lng, double zoom, double zerolat, double zerolng) {
            curLat = lat;
            curZoom = zoom;
            curLong = lng;
            curZeroLat = zerolat;
            curZeroLong = zerolng;
            System.out.println("Current long: " + lng);
            System.out.println("Current lat: " + lat);
            System.out.println("Current zoom: " + zoom);
            System.out.println("Current ZeroLat: " + zerolat);
            System.out.println("Current ZeroLong: " + zerolng);
            OSMMapEvent event = new OSMMapEvent(this, lat, lng, zoom);
    }

    public void setOnMapLatLngChanged(EventHandler<OSMMapEvent> eventHandler) {
        System.out.println("Current long: ");
        onMapLatLngZoomChanged = eventHandler;
    }

    public void setHeight(double h) {
        webView.setPrefHeight(h);
    }

    public void setWidth(double w) {
        webView.setPrefWidth(w);
    }

    public void updateLongLatZoom(){
        invokeJS("getLongLatZoom()");
    }

    public double getWidth(){
        return webView.getWidth();
    }

    public double getHeight(){
        return webView.getHeight();
    }
}

File: src/main/java/client/OSM/OSMMap.kt
package client.OSM

import javafx.beans.value.ObservableValue
import javafx.concurrent.Worker
import javafx.event.EventHandler
import javafx.scene.Parent
import javafx.scene.web.WebEngine
import javafx.scene.web.WebView
import netscape.javascript.JSObject

class OSMMap : Parent() {
    var lat00: Double = 0.0
    var lon00: Double = 0.0
    var lat11: Double = 0.0
    var lon11: Double = 0.0
    var curZoom = 15.0
    private var doc: JSObject? = null
    private var onMapLatLngZoomChanged: EventHandler<OSMMapEvent>? = null
    private var webView: WebView? = null
    private var ready = false
    private fun initMap() {
        webView = WebView()
        webView!!.engine.setJavaScriptEnabled(true)
        val path = "../../../../resources/main/evtTest.html"
        println(path)
        val resource = javaClass.getResource(path).toExternalForm()
        webView!!.engine.load(resource)
        children.add(webView)
        ready = false
        webView!!.engine.getLoadWorker().stateProperty()
            .addListener { observableValue: ObservableValue<out Worker.State>?, oldState: Worker.State?, newState: Worker.State ->
                if (newState == Worker.State.SUCCEEDED) {
                    ready = true
                }
            }
    }

    private fun initCommunication() {
        webView!!.engine.loadWorker.stateProperty()
            .addListener { observableValue: ObservableValue<out Worker.State>?, oldState: Worker.State?, newState: Worker.State ->
                if (newState == Worker.State.SUCCEEDED) {
                    doc = webView!!.engine.executeScript("window") as JSObject
                    doc!!.setMember("app", this@OSMMap)
                }
            }
    }

    private fun invokeJS(str: String) {
        webView!!.engine.executeScript(str)
        if (ready) {
            doc!!.eval(str)
        } else {
            webView!!.engine.loadWorker.stateProperty()
                .addListener { observableValue: ObservableValue<out Worker.State>?, oldState: Worker.State?, newState: Worker.State ->
                    if (newState == Worker.State.SUCCEEDED) {
                        doc!!.eval(str)
                    }
                }
        }
    }

    fun handle(
        lat00: Double,
        lon00: Double,
        lat11: Double,
        lon11: Double,
        zoom: Double
    ) {
        this.lat00 = lat00
        this.lon00 = lon00
        this.lat11 = lat11
        this.lon11 = lon11
        curZoom = zoom
    }

    fun setOnMapLatLngChanged(eventHandler: EventHandler<OSMMapEvent>?) {
        println("Current long: ")
        onMapLatLngZoomChanged = eventHandler
    }

    fun updateLongLatZoom() {
        invokeJS("getLongLatZoom()")
    }

    var width: Double
        get() = webView!!.width
        set(w) {
            webView!!.prefWidth = w
        }
    var height: Double
        get() = webView!!.height
        set(h) {
            webView!!.prefHeight = h
        }

    init {
        initMap()
        initCommunication()
    }
}